src/ImageRunModal.jsx

```diff
@@ -27,13 +27,112 @@ const units = {
     },
 };
 
+class PublishPortsForm extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            publish: [{ containerPort: null, hostPort: null, protocol: 'TCP' }],
+        };
+        this.removePort = this.removePort.bind(this);
+        this.addPort = this.addPort.bind(this);
+        this.onPortChange = this.onPortChange.bind(this);
+        this.validateAndChange = this.validateAndChange.bind(this);
+    }
+
+    removePort(idx, field, value) {
+        let ports = this.state.publish;
+        ports.splice(idx, 1);
+        if (ports.length === 0) {
+            this.addPort();
+        } else {
+            this.setState({ publish: ports });
+            this.validateAndChange(ports);
+        }
+    }
+
+    addPort() {
+        let ports = this.state.publish;
+        ports.push({ containerPort: null, hostPort: null, protocol: 'TCP' });
+        this.setState({ publish: ports });
+        this.validateAndChange(ports);
+    }
+
+    onPortChange(idx, field, value) {
+        let ports = this.state.publish;
+        ports[idx][field] = value || null;
+        this.setState({ publish: ports });
+        this.validateAndChange(ports);
+    }
+
+    validateAndChange(ports) {
+        if (typeof this.props.onChange !== 'function')
+            return;
+        let validated = ports.filter(port => parseInt(port.containerPort) >= 1 && parseInt(port.containerPort) <= 65535 &&
+                                     parseInt(port.hostPort) >= 1 && parseInt(port.hostPort) <= 65535);
+        this.props.onChange(validated.map(port => port.hostPort + ':' + port.containerPort + '/' + port.protocol));
+    }
+
+    render () {
+        const dialogValues = this.state;
+        return (
+            <div>
+                {
+                    dialogValues.publish.map((port, idx) => {
+                        return (
+                            <div className='publish-port-form ct-form-layout' key={'published-port-' + idx}>
```

Comment:
It feels a bit abuse-ish to use `ct-form-layout` for a non-`<form>`. Was this due to your struggles with layout and putting a table/form (port list) into a table/form (the whole dialog)?

Reply:
Yup, this is pretty much half a trial&error half @KKoukiou 's help. Didn't really manage to do it any other way, though there might be.

Reply:
We use ct-form-layout on non forms extensively, at least in machines. But maybe there is better solution indeed.